packages/core/src/magicswap/getSwapRoute.ts

```diff
@@ -0,0 +1,181 @@
+import type { NetworkInfo, RToken } from "@sushiswap/tines";
+import {
+  ConstantProductRPool,
+  findMultiRouteExactIn,
+  findMultiRouteExactOut,
+} from "@sushiswap/tines";
+import type { Config } from "@wagmi/core";
+import { parseUnits } from "viem";
+import type { AddressString, SupportedChainId } from "../types";
+import { multiplyArray, sumArray } from "../utils/array";
+import { bigIntToNumber } from "../utils/number";
+import { type Pool, fetchPools } from "./fetchPools";
+import type { PoolToken } from "./types";
+
+const tokenToRToken = ({
+  name,
+  symbol,
+  id: address,
+  decimals,
+}: PoolToken): RToken => ({
+  name,
+  symbol,
+  address,
+  decimals,
+});
+
+const createSwapRoute = (
+  tokenIn: PoolToken,
+  tokenOut: PoolToken | null,
+  pools: Pool[],
+  amount: bigint,
+  isExactOut: boolean,
+) => {
+  if (amount <= 0 || !tokenOut) {
+    return undefined;
+  }
+
+  const rTokenIn = tokenToRToken(tokenIn);
+  const rTokenOut = tokenToRToken(tokenOut);
+  const rPools = pools.map(
+    ({ id, token0, token1, reserve0, reserve1, totalFee }) => {
+      return new ConstantProductRPool(
+        id as AddressString,
+        tokenToRToken(token0),
+        tokenToRToken(token1),
+        Number(totalFee ?? 0),
+        parseUnits(reserve0.toString(), token0.decimals),
+        parseUnits(reserve1.toString(), token0.decimals),
+      );
+    },
+  );
+  const networks: NetworkInfo[] = [
+    {
+      baseToken: {
+        name: "ETH",
+        symbol: "ETH",
+        address: "0x0",
+        decimals: 18,
+      },
+      gasPrice: 0,
+    },
+  ];
+
+  if (isExactOut) {
+    return findMultiRouteExactOut(
+      rTokenIn,
+      rTokenOut,
+      amount,
+      rPools,
+      networks,
+    );
+  }
+
+  return findMultiRouteExactIn(rTokenIn, rTokenOut, amount, rPools, networks);
+};
+
+export const getSwapRoute = async ({
```

Comment:
Can you please break out `fetchPools` from this function's logic and instead have `pools` be a param that gets passed in? That allows us to make this function more reusable, makes the params clearer (e.g., no inventory or wagmi requirements) and makes it no longer `async`. Whoever wants to call this function should just call `fetchPools` first.